File: sparke-Java/src/main/java/com/SparkEngine/Recommendation.java
package com.SparkEngine;


import com.google.gson.Gson;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.mllib.recommendation.ALS;
import org.apache.spark.mllib.recommendation.MatrixFactorizationModel;
import org.apache.spark.mllib.recommendation.Rating;
import org.apache.spark.sql.Column;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Created by alikemal on 21.04.2017.
 */
public class Recommendation {
    JavaRDD<Rating> ratings;

    public Recommendation() {
        ratings = rating().toJavaRDD().map(row ->
                new Rating((int) row.getLong(4), (int) row.getLong(3), (int) row.getLong(1)));
    }

    public List<String> getSongs() {
        Dataset<Row> result = InitSpark.spark.sql("SELECT * FROM song limit 100");
        return result.toJSON().collectAsList();
    }

    public List<String> getSongbyTrackID(String trackID) {
        Dataset<Row> result = InitSpark.spark.sql("SELECT * FROM song limit 100").filter(new Column("trackID").equalTo(trackID));

        return result.toJSON().collectAsList();
    }

    public List<String> liststatnumber() {
        Dataset<Row> rs = InitSpark.spark.sql(
                "SELECT songID, row_number() OVER ( ORDER BY songID) as id" +
                        " FROM song group by songID ");

        return rs.toJSON().collectAsList();
    }

    public Dataset<Row> rating() {
        return InitSpark.spark.sql("SELECT * FROM ratingjson order by userID limit 200");
    }

    public String getRecommend() {
        final int rank = 5, iterations = 1, blocks = -1;
        ArrayList<String> result = new ArrayList<>();

        MatrixFactorizationModel model = ALS.train(ratings.rdd(), rank, iterations, 0.01, blocks);

        return new Gson().toJson(
                model.productFeatures().toJavaRDD().map(element ->
                        (element._1() + "," + Arrays.toString(element._2())))
                        .collect());
    }

}


/*
"SELECT u.id,s.id,ts.rating, tj.id " +
                "FROM tastestats ts" +
                "   INNER JOIN userjson u ON" +
                "   ts.userID= u.userID" +
                "   INNER JOIN songjson s ON" +
                "   ts.songID= s.songID "  +
                "   INNER JOIN song sg ON" +
                "   sg.songID= s.songID "  +
                "   INNER JOIN tag t ON" +
                "   sg.trackID= t.trackID "+
                "   INNER JOIN tagjson tj ON" +
                "   t.tagID= tj.tagID ");*/

// "SELECT tagID, row_number() OVER ( ORDER BY tagID) as id FROM tag group by tagID ");

/*
"SELECT *   " +
            "FROM tastestats st" +
            "   INNER JOIN song s ON" +
            "   st.songID= s.songID" +
            "   INNER JOIN tag t ON" +
            "   s.trackID= t.trackID " +
            "   limit 100"
 */

File: Recommendation.py
import json

from pyspark.mllib.recommendation import ALS

from InitSpark import InitSpark


class Recommendation:
    def __init__(self):
        self.init = InitSpark()
        self.spark = self.init.spark
        self.sc = self.init.sc

    def generateParquet(self) -> str:
        return self.init.generateParquet()

    def listPopulerSong(self) -> str:
        result = self.spark.sql("SELECT r.songid as product,0 as type FROM rating as r ORDER BY r.rating DESC LIMIT 20")

        return result.toJSON().collect().__str__()

    def listPopulerGenre(self, userid) -> str:
        ratingsRDD = self.spark.sql("SELECT * FROM rating").rdd \
            .map(lambda l: (int(l[3]), int(l[0]), float(l[1])))
        ratings = self.spark.createDataFrame(ratingsRDD)
        model = ALS.train(ratings, rank=5, iterations=5)
        recommend = model.recommendProducts(userid, 3)
        listRC = []
        for rate in recommend:
            listRC.append(int(rate[1]))
        df1 = self.spark.sql(
            "SELECT * FROM (SELECT r.genreID as genreID, r.songid as product ,2 as type,{0} as userid,rank() OVER (PARTITION BY r.genreID ORDER BY r.rating DESC) as rank FROM rating as r) s WHERE rank < 4".format(
                userid))

        result = df1.filter(
            (df1['genreID'] == listRC[0]) | (df1['genreID'] == listRC[1]) | (df1['genreID'] == listRC[2])).select(
            'product', 'type', 'userid')
        result.write.format("com.mongodb.spark.sql.DefaultSource") \
            .mode("append").save()
        return "ok"

    def ratingbyUserID(self, userid):
        ratingsRDD = self.spark.sql("SELECT * FROM rating").rdd \
            .map(lambda l: (int(l[3]), int(l[2]), float(l[1])))
        ratings = self.spark.createDataFrame(ratingsRDD)

        model = ALS.train(ratings, rank=5, iterations=5)
        recommend = model.recommendProducts(userid, 10)
        rmd = []
        for rate in recommend:
            rmd.append({"userid": userid, "product": rate.product, "type": 1})
        self.spark.createDataFrame(rmd).write.format("com.mongodb.spark.sql.DefaultSource") \
            .mode("append").save()
        return "ok"


        # "SELECT songID, row_number() OVER ( ORDER BY songID) as sarkiId"
        # ",songTitle as sarkiismi"
        # ",trackID"
        # ",artistName as sanatciIsmi"
        # ",genreId as genreId   " +
        # "FROM song "
